macros/src/lib.rs

```diff
@@ -40,22 +48,26 @@ pub fn contractimpl(metadata: TokenStream, input: TokenStream) -> TokenStream {
     let imp = parse_macro_input!(input as ItemImpl);
     let is_trait = imp.trait_.is_some();
     let ty = &imp.self_ty;
-    let derived = imp
-        .items
-        .iter()
-        .filter_map(|i| match i {
-            ImplItem::Method(m) => Some(m),
-            _ => None,
-        })
+    let derived = get_methods(&imp)
         .filter(|m| is_trait || matches!(m.vis, Visibility::Public(_)))
         .map(|m| {
             let ident = &m.sig.ident;
-            let call = quote! { <#ty>::#ident };
-            derive_fn(&call, ident, &m.sig.inputs, &m.sig.output, &args.feature)
+            let call = quote! { <super::#ty>::#ident };
+            let trait_ident = imp.trait_.as_ref().and_then(|x| x.1.get_ident());
+            derive_fn(
+                &call,
+                ident,
+                &m.sig.inputs,
+                &m.sig.output,
+                &args.export_if,
+                &trait_ident,
+            )
         });
+    let add_functions = derive_add_functions(ty, get_methods(&imp), &args.tests_if);
```

Comment:
This is a smart way to introduce stateless function registration. However, I suspect it will break contracts that have multiple `impl` blocks, since it will implement the `AddFunctions` trait multiple times. Do you know if that will work?

I suspect we need a different approach where we attempt to produce stateful storage, but if what you have in this PR works for at least some contracts, it is probably a good place for us to start and we can create a follow up issue for supporting multiple impl blocks.

Reply:
Tracked in https://github.com/stellar/rs-stellar-contract-sdk/issues/268